modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cluster/ClientClusterGroup.cs

```diff
@@ -0,0 +1,311 @@
+    using Apache.Ignite.Core.Impl.Binary.IO;
+    using System.Linq;
+    using Apache.Ignite.Core.Impl.Common;
+
+    /// <summary>
+    /// Ignite client projection implementation.
+    /// </summary>
+    internal class ClientClusterGroup : IClientClusterGroup
+    {
+        /** Attribute: platform. */
+        private const string AttrPlatform = "org.apache.ignite.platform";
+
+        /** Platform. */
+        private const string Platform = "dotnet";
+
+        /**
+         * Initial topology version; invalid from Java perspective,
+         * so update will be triggered when this value is met.
+         */
+        private const int TopVerInit = 0;
+
+        /** Ignite. */
+        private readonly IgniteClient _ignite;
+
+        /** Marshaller. */
+        private readonly Marshaller _marsh;
+
+        /** Topology version. */
+        private long _topVer = TopVerInit;
+
+        /** Current projection. */
+        private readonly ClientClusterGroupProjection _projection;
+
+        /** Predicate. */
+        private readonly Func<IClusterNode, bool> _predicate;
+
+        /** Node ids collection. */
+        private IList<Guid> _nodeIds;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="ignite">Ignite.</param>
+        /// <param name="marsh">Marshaller.</param>
+        internal ClientClusterGroup(IgniteClient ignite, Marshaller marsh)
+            : this(ignite, marsh, ClientClusterGroupProjection.Empty)
+        {
+        }
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="ignite">Ignite.</param>
+        /// <param name="marsh">Marshaller.</param>
+        /// <param name="projection">Projection.</param>
+        /// <param name="predicate">Predicate.</param>
+        private ClientClusterGroup(IgniteClient ignite, Marshaller marsh,
+            ClientClusterGroupProjection projection, Func<IClusterNode, bool> predicate = null)
+        {
+            Debug.Assert(ignite != null);
+            Debug.Assert(marsh != null);
+
+            _ignite = ignite;
+            _marsh = marsh;
+            _projection = projection;
+            _predicate = predicate;
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForAttribute(string name, string val)
+        {
+            IgniteArgumentCheck.NotNullOrEmpty(name, "name");
+
+            return new ClientClusterGroup(_ignite, _marsh, _projection.ForAttribute(name, val));
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForDotNet()
+        {
+            return ForAttribute(AttrPlatform, Platform);
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForServers()
+        {
+            return new ClientClusterGroup(_ignite, _marsh, _projection.ForServerNodes());
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForPredicate(Func<IClusterNode, bool> p)
```

Comment:
Let's remove this method - LINQ is there to filter by predicate locally.

Reply:
Or maybe I'm wrong, this provides an easy way to have filtered `IClientClusterGroup` instance, let's keep it.